examples/langchain/simple_agent.py

```diff
@@ -0,0 +1,33 @@
+import os
+import mlflow
+
+from langchain.agents import load_tools
+from langchain.agents import initialize_agent
+from langchain.agents import AgentType
+from langchain.llms import OpenAI
+
+assert "OPENAI_API_KEY" in os.environ, "Please set the OPENAI_API_KEY environment variable."
+assert "SERPAPI_API_KEY" in os.environ, "Please set the SERPAPI_API_KEY environment variable."
+
+# First, let's load the language model we're going to use to control the agent.
+llm = OpenAI(temperature=0)
+
+# Next, let's load some tools to use. Note that the `llm-math` tool uses an LLM, so we need to pass that in.
+tools = load_tools(["serpapi", "llm-math"], llm=llm)
+
+# Finally, let's initialize an agent with the tools, the language model, and the type of agent we want to use.
+agent = initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True)
+
+with mlflow.start_run():
+    logged_model = mlflow.langchain.log_model(agent, "langchain_model")
```

Comment:
This seems to be failing for me on `langchain==0.0.132` and `langchain==0.0.148` (latest):

```
  File "simple_agent.py", line 22, in <module>
    logged_model = mlflow.langchain.log_model(agent, "langchain_model")
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 323, in log_model
    return Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 547, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 172, in save_model
    model_data_kwargs = _save_model(lc_model, path)
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 353, in _save_model
    model.save_agent(agent_data_path)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/agent.py", line 599, in save_agent
    return self.agent.save(file_path)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/agent.py", line 145, in save
    agent_dict = self.dict()
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/agent.py", line 119, in dict
    _dict = super().dict()
  File "pydantic/main.py", line 445, in pydantic.main.BaseModel.dict
  File "pydantic/main.py", line 843, in _iter
  File "pydantic/main.py", line 718, in pydantic.main.BaseModel._get_value
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/schema.py", line 391, in dict
    output_parser_dict["_type"] = self._type
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/schema.py", line 386, in _type
    raise NotImplementedError
NotImplementedError
```

Am I doing something incorrectly here?

Reply:
Is https://github.com/hwchase17/langchain/issues/3217 related?

Reply:
@sunishsheth2009 @harupy This still fails in langchain versions < 0.0.142 with the following:

```
2023/04/25 14:44:10 WARNING mlflow.utils.environment: Encountered an unexpected error while inferring pip requirements (model URI: /var/folders/v3/fwmb62ds0bd7700x2vgm051r0000gp/T/tmp43tgsosa/model, flavor: langchain), fall back to return ['langchain==0.0.141']. Set logging level to DEBUG to see the full traceback.
2023/04/25 14:44:10 WARNING mlflow.pyfunc: Detected one or more mismatches between the model's dependencies and the current Python environment:
 - mlflow (current: 2.1.2.dev0, required: mlflow<3,>=2.3)
To fix the mismatches, call `mlflow.pyfunc.get_model_dependencies(model_uri)` to fetch the model's environment and install dependencies using the resulting environment file.
Traceback (most recent call last):
  File "simple_agent.py", line 24, in <module>
    loaded_model = mlflow.pyfunc.load_model(logged_model.model_uri)
  File "/Users/corey.zumar/mlflow/mlflow/pyfunc/__init__.py", line 596, in load_model
    model_impl = importlib.import_module(conf[MAIN])._load_pyfunc(data_path)
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 451, in _load_pyfunc
    return wrapper_cls(_load_model_from_local_fs(path))
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 471, in _load_model_from_local_fs
    return _load_model(lc_model_path, agent_model_path, tools_model_path, agent_primitive_path)
  File "/Users/corey.zumar/mlflow/mlflow/langchain/__init__.py", line 409, in _load_model
    model = initialize_agent(tools=tools, llm=llm, agent_path=agent_path, **kwargs)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/initialize.py", line 56, in initialize_agent
    agent_obj = load_agent(
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/loading.py", line 94, in load_agent
    return _load_agent_from_file(path, **kwargs)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/langchain/agents/loading.py", line 112, in _load_agent_from_file
    config = yaml.safe_load(f)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/__init__.py", line 125, in safe_load
    return load(stream, SafeLoader)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 413, in construct_yaml_map
    value = self.construct_mapping(node)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 218, in construct_mapping
    return super().construct_mapping(node, deep=deep)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 143, in construct_mapping
    value = self.construct_object(value_node, deep=deep)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 100, in construct_object
    data = constructor(self, node)
  File "/Users/corey.zumar/opt/anaconda3/envs/base2/lib/python3.8/site-packages/yaml/constructor.py", line 427, in construct_undefined
    raise ConstructorError(None, None,
yaml.constructor.ConstructorError: could not determine a constructor for the tag 'tag:yaml.org,2002:python/object/apply:langchain.agents.agent_types.AgentType'
  in "/Users/corey.zumar/mlflow/examples/langchain/mlruns/0/624f4a4836074b5b9ad8cd73dd315381/artifacts/langchain_model/agent.yaml", line 1, column 8
```

Is there a version that you found passes?

Reply:
Yes that is resolved here: https://github.com/hwchase17/langchain/pull/2998 :)